pkg/identity/gitlabcom/principal_test.go

```diff
@@ -0,0 +1,282 @@
+				url:               "https://gitlab.com/",
+				eventName:         "push",
+				pipelineID:        "757451528",
+				repository:        "cpanato/testing-cosign",
+				repositoryID:      "42831435",
+				repositoryOwner:   "cpanato",
+				repositoryOwnerID: "1730270",
+				jobID:             "3659681386",
+				ref:               "refs/heads/main",
+				runnerID:          1,
+				runnerEnvironment: "gitlab-hosted",
+				sha:               "714a629c0b401fdce83e847fc9589983fc6f46bc",
+			},
+			WantErr: false,
+		},
+		`Token missing pipeline_source claim should be rejected`: {
+			Claims: map[string]interface{}{
+				"aud":            "sigstore",
+				"exp":            0,
+				"iss":            "https://gitlab.com",
+				"sub":            "project_path:cpanato/testing-cosign:ref_type:branch:ref:main",
+				"project_id":     "42831435",
+				"project_path":   "cpanato/testing-cosign",
+				"namespace_path": "cpanato",
+				"namespace_id":   "1730270",
+				"pipeline_id":    "757451528",
+				"job_id":         "3659681386",
+				"ref":            "main",
+				"ref_type":       "branch",
+			},
+			WantErr:     true,
+			ErrContains: "pipeline_source",
+		},
+		`Token missing project_path claim should be rejected`: {
+			Claims: map[string]interface{}{
+				"aud":             "sigstore",
+				"exp":             0,
+				"iss":             "https://gitlab.com",
+				"sub":             "project_path:cpanato/testing-cosign:ref_type:branch:ref:main",
+				"project_id":      "42831435",
+				"pipeline_id":     "757451528",
+				"namespace_id":    "1730270",
+				"pipeline_source": "push",
+				"namespace_path":  "cpanato",
+				"job_id":          "3659681386",
+				"ref":             "main",
+				"ref_type":        "branch",
+			},
+			WantErr:     true,
+			ErrContains: "project_path",
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			token := &oidc.IDToken{
+				Issuer:  test.Claims["iss"].(string),
+				Subject: test.Claims["sub"].(string),
+			}
+			claims, err := json.Marshal(test.Claims)
+			if err != nil {
+				t.Fatal(err)
+			}
+			withClaims(token, claims)
+
+			untyped, err := JobPrincipalFromIDToken(context.TODO(), token)
+			if err != nil {
+				if !test.WantErr {
+					t.Fatal("didn't expect error", err)
+				}
+				if !strings.Contains(err.Error(), test.ErrContains) {
+					t.Fatalf("expected error %s to contain %s", err, test.ErrContains)
+				}
+				return
+			}
+			if err == nil && test.WantErr {
+				t.Fatal("expected error but got none")
+			}
+
+			principal, ok := untyped.(*jobPrincipal)
+			if !ok {
+				t.Errorf("Got wrong principal type %v", untyped)
+			}
+			if *principal != test.ExpectPrincipal {
+				t.Errorf("got %v principal and expected %v", *principal, test.ExpectPrincipal)
+			}
+		})
+	}
+}
+
+// reflect hack because "claims" field is unexported by oidc IDToken
+// https://github.com/coreos/go-oidc/pull/329
+func withClaims(token *oidc.IDToken, data []byte) {
```

Comment:
Out of scope for this PR, but we should look into pulling this out into a library so we're not copying the same hacks for each identity provider. 🤔

Reply:
sounds good, i can work on that in a follow up